src/core/worker.js

```diff
@@ -621,15 +623,22 @@ class WorkerMessageHandler {
             }
           }
 
-          const xfa = (acroForm instanceof Dict && acroForm.get("XFA")) || [];
+          const xfa = (acroForm instanceof Dict && acroForm.get("XFA")) || null;
           let xfaDatasets = null;
+          let hasDatasets = false;
```

Comment:
Do you actually need this separate boolean?
Can't you just check `!!xfaDatasets` in the `src/core/writer.js` code, and thus avoid the need to track closely related state twice?

Reply:
In case `xfaDatasets` is `null`, I set it to a new reference after having visited the `XFA` array so it's no more null.
So the only way for `xfaDatasets` to be `null` is to not have an array for `XFA`.
So when we have an array we must know if the reference we've has been newly created (and in this case we must update AcroForm) or not.

We could probably get rid of this boolean in moving https://github.com/mozilla/pdf.js/blob/master/src/core/worker.js#L650 in `writer.js` because the afaik the xref for the xref table must be the last one (so it's why I get the new ref for `xfaDatasets` before calling `incrementalUpdate`), but I guess it's better to do that in a follow-up.